### PR TITLE
Store GADT constraints until PostTyper

### PR DESCRIPTION
**Pull request by abgruszecki**
In some cases, the attachment can get dropped. We have a lot of such cases. Most seem to happen for tuple patterns.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 1]
```diff
@@ -66,6 +66,9 @@ object Typer {
     if (!tree.isEmpty && !tree.isInstanceOf[untpd.TypedSplice] && ctx.typerState.isGlobalCommittable)
       assert(tree.span.exists, i"position not set for $tree # ${tree.uniqueId} of ${tree.getClass} in ${tree.source}")
 
+  /** An attachment for GADT constraints that were inferred for a pattern. */
+  val InferredGadtConstraints = new Property.Key[core.GadtConstraint]
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala:70]: This should be a StickyKey to not get lost through tree transformations:
```suggestion
  val InferredGadtConstraints = new Property.StickyKey[core.GadtConstraint]
```

File: compiler/src/dotty/tools/dotc/transform/PostTyper.scala [hunk 2]
```diff
@@ -257,6 +257,14 @@ class PostTyper extends MacroTransform with IdentityDenotTransformer { thisPhase
 
     override def transform(tree: Tree)(using Context): Tree =
       try tree match {
+        case CaseDef(pat, _, _) =>
+          val gadtCtx =
+           pat.getAttachment(typer.Typer.InferredGadtConstraints) match
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/PostTyper.scala:262]: If they're not used afterwards we don't want to keep them around:
```suggestion
           pat.removeAttachment(typer.Typer.InferredGadtConstraints) match
```

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 3]
```diff
@@ -1598,9 +1601,12 @@ class Typer extends Namer
       val pat1 = indexPattern(tree).transform(pat)
       val guard1 = typedExpr(tree.guard, defn.BooleanType)
       var body1 = ensureNoLocalRefs(typedExpr(tree.body, pt1), pt1, ctx.scope.toList)
+      pat1.putAttachment(InferredGadtConstraints, ctx.gadt)
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala:1604]: We should only do this if the gadt constraint is non-empty

**Comment by odersky** [comment]
test performance please

**Review by odersky — CHANGES_REQUESTED** [review]
I am wary that this could give bad memory leaks. GADT constraints can be large.

Why do we need GADT constraints in PostTyper? The PR does not explain, nor does the code.

**Comment by abgruszecki** [comment]
@odersky there's a serious problem - type application that needs GADT constraints simply doesn't work. https://github.com/lampepfl/dotty/issues/12226 is one example. Another one is that without this PR we get an error on this line: https://github.com/lampepfl/dotty/pull/12258/files#diff-953aaf705b431c6945ad2bd63ce89e017169f7373c45f9c40f265f4d11705a81R11.

The reason why type application doesn't work is that we check it in PostTyper, when we already forgot the GADT constraints. I tried moving the checks to Typer, but it seems like there's no clean way to do that, for instance b/c when we exit `typedTypeApply`, the type arguments may still be variables. So that means we should either re-infer the constraints in PostTyper, or store and extract them. Of those two solutions, storing the constraints is by far the cleanest and the most future-proof.

If we ever get problems because of too many GADT constraints, we could simply infer less of them. It's always sound to infer less GADT constraints. IIRC, GHC also stores its GADT constraints, and won't record constraints past some threshold.

**Comment by odersky** [comment]
OK. I'd just add a comment explaining why we need to attach GADT constraints

**Review by smarter — APPROVED** [review]

**Pull request merged**